In [110]:
#Import libraries
import pandas as pd
import numpy as np

# DESCRIPTOR DATASETS
Load to get names of columns

In [111]:
#Load cells geo descriptor
df_cdr = pd.read_excel('cells_geo_description.xlsx')
df_cdr

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

In [112]:
#Load Telcom datasets descriptors
df_cdr = pd.read_excel('CDR_description.xlsx')
df_cdr

Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

# Cells Geo
Loading and cleaning the cells geo dataframe

In [243]:
#Loading cells geo dataset as a panda dataframe
# Load the csv file into a data frame
df_cells = pd.read_csv('/content/cells_geo.csv', delimiter=';',index_col=0)

#  Preview of first 5 rows
df_cells.head()

VILLES      STATUS LOCALISATION  ... AREA     CELL_ID   SITE_CODE
0   ADJAME  In Service      ABIDJAN  ...  NaN  fbb2e75af3  0bb492ef17
1   ADJAME  In Service      ABIDJAN  ...  NaN  1502501d78  0bb492ef17
2   ADJAME  In Service      ABIDJAN  ...  NaN  0c2e182a0c  0bb492ef17
3  ASSINIE  In Service    INTERIEUR  ...  NaN  69bdfc69fd  8a1e6ce31f
4  ASSINIE  In Service    INTERIEUR  ...  NaN  7a1029d2ab  8a1e6ce31f

[5 rows x 11 columns]

In [244]:
#Aligning column names with the cells geo descriptor and CDR descriptor

columns= ['VILLES','STATUS','LOCALISATION','DECOUPZONE','ZONENAME','LONGITUDE','LATITUDE','REGION','AREA','CELL_ID','SITE_CODE']
df_cells.columns = columns
df_cells.columns

Index(['VILLES', 'STATUS', 'LOCALISATION', 'DECOUPZONE', 'ZONENAME',
       'LONGITUDE', 'LATITUDE', 'REGION', 'AREA', 'CELL_ID', 'SITE_CODE'],
      dtype='object')

In [248]:
#Checking rows and columns of cells geo
df_cells.shape

(3974, 6)

In [245]:
#Renaming columns to align with the other dataset
df_cells.rename(columns={'VILLES':'CITY','SITE_CODE':'SITE_ID'}, inplace=True)
df_cells.columns

Index(['CITY', 'STATUS', 'LOCALISATION', 'DECOUPZONE', 'ZONENAME', 'LONGITUDE',
       'LATITUDE', 'REGION', 'AREA', 'CELL_ID', 'SITE_ID'],
      dtype='object')

In [246]:
# Analysis is on Cities and regions the other columns will not be necessary when compared to telcom datasets
df_cells.drop(['DECOUPZONE', 'ZONENAME', 'LONGITUDE','LATITUDE', 'AREA'], axis=1, inplace=True)

In [247]:
#check for null values
df_cells.isnull().sum()

CITY             0
STATUS          67
LOCALISATION     0
REGION           0
CELL_ID          0
SITE_ID          0
dtype: int64

In [251]:
#Drop null values
df_cells.dropna(inplace=True)


In [252]:
#check
df_cells.isnull().sum()

CITY            0
STATUS          0
LOCALISATION    0
REGION          0
CELL_ID         0
SITE_ID         0
dtype: int64

In [254]:
df_cells.head()

CITY      STATUS LOCALISATION     REGION     CELL_ID     SITE_ID
0   ADJAME  In Service      ABIDJAN     LAGUNE  fbb2e75af3  0bb492ef17
1   ADJAME  In Service      ABIDJAN     LAGUNE  1502501d78  0bb492ef17
2   ADJAME  In Service      ABIDJAN     LAGUNE  0c2e182a0c  0bb492ef17
3  ASSINIE  In Service    INTERIEUR  SUD COMOE  69bdfc69fd  8a1e6ce31f
4  ASSINIE  In Service    INTERIEUR  SUD COMOE  7a1029d2ab  8a1e6ce31f

# Telcom 
Loading and cleaning the Telcom dataframe

In [256]:
#Load Telcom dataset
df_t1=pd.read_csv('/content/Telcom_dataset.csv')
#Preview first 5 rows
df_t1.head()

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [257]:
#Aligning column names with CDR description

columnsCDR=['PRODUCT','VALUE','DATE_TIME','CELL_ON_SITE','DW_A_NUMBER_INT','DW_B_NUMBER_INT','COUNTRY_A','COUNTRY_B','CELL_ID','SITE_ID']
df_t1.columns=columnsCDR
#Checking
df_t1.columns

Index(['PRODUCT', 'VALUE', 'DATE_TIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT',
       'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELL_ID', 'SITE_ID'],
      dtype='object')

In [261]:
#Checking rows and columns of Telcom dataset
df_t1.shape

(5001, 5)

In [259]:
#based on the dataset description, deleting columns that will not be needed
df_t1.drop(['CELL_ON_SITE','DW_A_NUMBER_INT','DW_B_NUMBER_INT','COUNTRY_A','COUNTRY_B'], axis=1, inplace=True)

In [260]:
#checking for null values
df_t1.isnull().sum()

PRODUCT        0
VALUE          0
DATE_TIME      0
CELL_ID        0
SITE_ID      716
dtype: int64

In [262]:
#Drop the null values 
df_t1.dropna(inplace=True)

In [263]:
#check missing values are dropped
df_t1.isnull().sum()

PRODUCT      0
VALUE        0
DATE_TIME    0
CELL_ID      0
SITE_ID      0
dtype: int64

# Telcom dataset 2
Loading and cleaning Telcom 2 dataframe

In [264]:
#Load Telcom 2 dataset
df_t2=pd.read_csv('/content/Telcom_dataset2.csv')
#Preview first 5 rows
df_t2.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO

[5 rows x 10 columns]

In [265]:
#Aligning column names with CDR description

columnsCDR=['PRODUCT','VALUE','DATE_TIME','CELL_ON_SITE','DW_A_NUMBER_INT','DW_B_NUMBER_INT','COUNTRY_A','COUNTRY_B','CELL_ID','SITE_ID']
df_t2.columns=columnsCDR
#Checking
df_t2.columns

Index(['PRODUCT', 'VALUE', 'DATE_TIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT',
       'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELL_ID', 'SITE_ID'],
      dtype='object')

In [266]:
#Checking rows and columns of Telcom dataset
df_t2.shape

(5001, 10)

In [267]:
#based on the dataset description, deleting columns that will not be needed
df_t2.drop(['CELL_ON_SITE','DW_A_NUMBER_INT','DW_B_NUMBER_INT','COUNTRY_A','COUNTRY_B'], axis=1, inplace=True)

In [268]:
#checking for null values
df_t2.isnull().sum()

PRODUCT        0
VALUE          0
DATE_TIME      0
CELL_ID        0
SITE_ID      676
dtype: int64

In [269]:
#Drop the null values since its relatively a small number
df_t2.dropna(inplace=True)

In [270]:
#check missing values dropped
df_t2.isnull().sum()

PRODUCT      0
VALUE        0
DATE_TIME    0
CELL_ID      0
SITE_ID      0
dtype: int64

# Telcom Dataset 3
Loading and cleaning dataset 3

In [271]:
#Load Telcom dataset
df_t3=pd.read_csv('/content/Telcom_dataset3.csv')
#Preview first 5 rows
df_t3.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

In [272]:
#Aligning column names with CDR description

columnsCDR=['PRODUCT','VALUE','DATE_TIME','CELL_ON_SITE','DW_A_NUMBER_INT','DW_B_NUMBER_INT','COUNTRY_A','COUNTRY_B','CELL_ID','SITE_ID']
df_t3.columns=columnsCDR
#Checking
df_t3.columns

Index(['PRODUCT', 'VALUE', 'DATE_TIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT',
       'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELL_ID', 'SITE_ID'],
      dtype='object')

In [273]:
#Checking rows and columns of Telcom dataset
df_t3.shape

(5001, 10)

In [274]:
#based on the dataset description, deleting columns that will not be needed
df_t3.drop(['CELL_ON_SITE','DW_A_NUMBER_INT','DW_B_NUMBER_INT','COUNTRY_A','COUNTRY_B'], axis=1, inplace=True)

In [275]:
#checking for null values
df_t3.isnull().sum()

PRODUCT        0
VALUE          0
DATE_TIME      0
CELL_ID        0
SITE_ID      605
dtype: int64

In [277]:
#Drop the null values its relatively a small number
df_t3.dropna(inplace=True)

In [278]:
#check missing values dropped
df_t3.isnull().sum()

PRODUCT      0
VALUE        0
DATE_TIME    0
CELL_ID      0
SITE_ID      0
dtype: int64

# Telcom Datasets
Merging and cleaning the three telcom datasets


In [449]:
#concating the 3 telcom datasets
df_all= pd.concat([df_t1,df_t2,df_t3],ignore_index=True)
#showing a priview
df_all.head()

PRODUCT  VALUE              DATE_TIME     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  051KhYDCpv  051KhYDCpv

In [280]:
#Check the Rows and columns
df_all.shape

(13006, 5)

In [410]:
#Checking for null values
df_all.isnull().sum()

PRODUCT      0
VALUE        0
DATE_TIME    0
CELL_ID      0
SITE_ID      0
dtype: int64

In [282]:
#Checking datatypes
df_all.dtypes

PRODUCT      object
VALUE         int64
DATE_TIME    object
CELL_ID      object
SITE_ID      object
dtype: object

In [450]:
#Changing Date_Time from object data type to datetime datatype
df_all['DATE_TIME']=pd.to_datetime(df_all['DATE_TIME'])

In [284]:
#confirm data type changes
df_all.dtypes

PRODUCT              object
VALUE                 int64
DATE_TIME    datetime64[ns]
CELL_ID              object
SITE_ID              object
dtype: object

In [451]:
#To split date-time to date column and time column
#To get date column
df_all['date'] = [d.date() for d in df_all['DATE_TIME']]
#To get time column
df_all['time'] = [d.time() for d in df_all['DATE_TIME']]
df_all


PRODUCT  VALUE           DATE_TIME  ...     SITE_ID        date      time
0       Voice      0 2012-05-06 23:04:37  ...  /TJNe+Mmtu  2012-05-06  23:04:37
1         sms      0 2012-05-06 23:05:25  ...  /+cKZKIp41  2012-05-06  23:05:25
2         sms     25 2012-05-06 23:05:41  ...  /u0FSD+ahi  2012-05-06  23:05:41
3       Voice     56 2012-05-06 23:05:42  ...  +cNeJzsTp3  2012-05-06  23:05:42
4       Voice     14 2012-05-06 23:06:29  ...  051KhYDCpv  2012-05-06  23:06:29
...       ...    ...                 ...  ...         ...         ...       ...
13001   Voice      0 2012-05-09 00:01:43  ...  f7d63a29e5  2012-05-09  00:01:43
13002   Voice     62 2012-05-09 00:01:43  ...  b4ddb9e93c  2012-05-09  00:01:43
13003   Voice     32 2012-05-09 00:01:43  ...  d68260ef50  2012-05-09  00:01:43
13004   Voice     30 2012-05-09 00:01:43  ...  b4130b54ab  2012-05-09  00:01:43
13005   Voice     60 2012-05-09 00:01:43  ...  6d70dae094  2012-05-09  00:01:43

[13006 rows x 7 columns]

# Merging the Telcom datasets to Cells Geo
Merging the concated dataframe of telcom, telcom2 and telcom 3 to dataframe cells geo and cleaning

In [473]:
merge = pd.merge(df_cells, df_all,on='SITE_ID')
merge.head()

CITY      STATUS LOCALISATION  ...   CELL_ID_y        date      time
0  ADJAME  In Service      ABIDJAN  ...  ac8f1ccb39  2012-05-07  00:00:58
1  ADJAME  In Service      ABIDJAN  ...  ac8f1ccb39  2012-05-07  00:00:58
2  ADJAME  In Service      ABIDJAN  ...  ac8f1ccb39  2012-05-07  00:00:58
3   ABOBO  In Service      ABIDJAN  ...  35a513722e  2012-05-07  00:00:18
4   ABOBO  In Service      ABIDJAN  ...  35a513722e  2012-05-07  00:00:30

[5 rows x 12 columns]

In [469]:
#checking the shape of the new dataset
merge.shape

(50645, 12)

the new dataset has 50,645 rows and 12 columns

In [321]:
#Check null values
merge.isnull().sum()

CITY            0
STATUS          0
LOCALISATION    0
REGION          0
CELL_ID_x       0
SITE_ID         0
PRODUCT         0
VALUE           0
DATE_TIME       0
CELL_ID_y       0
Date            0
Time            0
dtype: int64

# Analysis

In [475]:
#checking for product usage by date
merge.groupby(['date','PRODUCT'])['PRODUCT'].count().sort_values(ascending=False)


date        PRODUCT
2012-05-09  sms        12262
2012-05-08  sms         8967
2012-05-07  Voice       8353
            sms         8024
2012-05-08  Voice       7123
2012-05-09  Voice       5916
Name: PRODUCT, dtype: int64

The product voice had the most use on 7/5/2012 compared to sms which had the most usage on 9/5/2012

In [331]:
#checking for city usage for MTN PRODUCTS
merge.groupby(['PRODUCT','CITY'])['PRODUCT'].count().sort_values(ascending=False).head(10)

PRODUCT  CITY     
sms      YOPOUGON     5560
         ABOBO        4260
Voice    YOPOUGON     2980
sms      COCODY       2105
Voice    ABOBO        1935
         COCODY       1709
sms      ADJAME       1533
         KOUMASSI     1369
         ATTECOUBE    1206
         MARCORY      1205
Name: PRODUCT, dtype: int64

The top three cities for the sms product are 
    

*   YOPOUGON 5560
*   ABOBO 4260
*   COCODY       2105

Top three cities for voice
* YOPOUGON     2980   
* ABOBO        1935 
* COCODY       1709

In [332]:
#Checking city least MTN product usage 
merge.groupby(['PRODUCT','CITY'])['PRODUCT'].count().sort_values(ascending=True).head(10)

PRODUCT  CITY         
Voice    KOTOULA          2
         DANANON          2
sms      BIANKOUMA        3
Voice    GUIBEROUA        3
sms      TAFIRE           3
Voice    GUINTEGUELA      3
sms      BECEDIBRIGNAN    3
         BACANDA          3
         AYAPRIKRO        3
         AYAME            3
Name: PRODUCT, dtype: int64

The bottom three cities for the sms product are 
    

*   BIANKOUMA        3
*   TAFIRE           3
*   BECEDIBRIGNAN    **3**

Bottom three cities for voice
* KOTOULA          2
* DANANON          2 
* GUIBEROUA        3

In [342]:
#Checking for most popular region for MTN products
merge.groupby(['PRODUCT','REGION'])['PRODUCT'].count().sort_values(ascending=False).head(10)

PRODUCT  REGION           
sms      LAGUNE               20779
Voice    LAGUNE               12509
sms      BAS SASSANDRA         1304
         HAUT SASSANDRA        1281
         LACS                  1264
Voice    LACS                  1220
         BAS SASSANDRA         1091
         HAUT SASSANDRA        1020
         VALLEE DU BANDAMA      828
sms      SUD BANDAMA            768
Name: PRODUCT, dtype: int64

In [343]:
#Checking for least popular region for MTN products
merge.groupby(['PRODUCT','REGION'])['PRODUCT'].count().sort_values(ascending=True).head(10)

PRODUCT  REGION       
sms      BAFING            36
Voice    BAFING            66
sms      ZANZAN            93
         SUD COMOE         93
Voice    SUD COMOE        111
sms      MOYEN COMOE      135
         DENGUELE         147
Voice    MOYEN COMOE      186
sms      NZI COMOE        206
         MOYEN CAVALLY    225
Name: PRODUCT, dtype: int64

In [476]:
#Checking the product is the most used among users
merge.groupby(['PRODUCT'])['PRODUCT'].count().sort_values(ascending=True)

PRODUCT
Voice    21392
sms      29253
Name: PRODUCT, dtype: int64